storage/src/global_state/state/mod.rs

```diff
@@ -1885,27 +1887,94 @@ pub trait StateProvider: Send + Sync + Sized {
+                                        )),
+                                    },
+                                    None => {
+                                        EntryPointResult::ValueNotFound(query_result_not_found_msg)
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    QueryResult::Failure(tce) => EntryPointResult::Failure(tce),
+                    QueryResult::Success { value, .. } => {
+                        if let StoredValue::EntryPoint(entry_point) = *value {
+                            EntryPointResult::Success { entry_point }
+                        } else {
+                            error!("Expected to get entry point value received other variant");
+                            EntryPointResult::Failure(
+                                TrackingCopyError::UnexpectedStoredValueVariant,
+                            )
+                        }
+                    }
+                }
+            }
+            Err(_) => EntryPointResult::Failure(
+                //TODO maybe we can have a better error type here
+                TrackingCopyError::ValueNotFound("Entry point not found".to_string()),
+            ),
+        }
+    }
+
+    /// Gets an contract value.
+    fn contract(&self, request: ContractRequest) -> ContractResult {
+        let query_request = QueryRequest::new(request.state_hash(), request.key(), vec![]);
 
         match self.query(query_request) {
-            QueryResult::RootNotFound => EntryPointsResult::RootNotFound,
-            QueryResult::ValueNotFound(msg) => EntryPointsResult::ValueNotFound(msg),
-            QueryResult::Failure(tce) => EntryPointsResult::Failure(tce),
+            QueryResult::RootNotFound => ContractResult::RootNotFound,
+            QueryResult::ValueNotFound(msg) => ContractResult::ValueNotFound(msg),
+            QueryResult::Failure(tce) => ContractResult::Failure(tce),
             QueryResult::Success { value, .. } => {
-                if let StoredValue::EntryPoint(entry_point_value) = *value {
-                    EntryPointsResult::Success {
-                        entry_point: entry_point_value,
-                    }
+                if let StoredValue::Contract(contract) = *value {
+                    ContractResult::Success { contract }
                 } else {
-                    error!("Expected to get entry point value received other variant");
-                    EntryPointsResult::Failure(TrackingCopyError::UnexpectedStoredValueVariant)
+                    error!("Expected to get contract value received other variant");
+                    ContractResult::Failure(TrackingCopyError::UnexpectedStoredValueVariant)
                 }
             }
         }
     }
 
+    /// Gets an entry point value.
+    fn entry_point_exists(&self, request: EntryPointsRequest) -> EntryPointExistsResult {
```

Comment:
I acknowledge that it does work to re-use EntryPointsRequest here, but to the best of my knowledge we have kept 1:1 parity between request and response pairs (though I could be wrong). Thus I would think we would doe one of the following: 

1. have an EntryPointExistsRequest corresponding to EntryPointExistsResponse to adhere to the request / response isomorphism convention.
2. instead of adding EntryPointExistsResponse, add a variant to EntryPointResponse to handle the exists result, and modify EntryPointRequest to have a bool existence_check indicating the caller only wants t/f and does not want the entry point instance.